File: ts/src/overnight-first-customer-v2.ts
/**
 * Strict TypeScript/Zod mirror of Python overnight_first_customer_v2.py.
 * Python remains authoritative; parity constants live in the adjacent test.
 */
import { createHash } from 'node:crypto';
import { z } from 'zod';

import { DigestSchema } from './factory/planet-output.js';
import { sha256Digest } from './factory/digest.js';

export const FIRST_CUSTOMER_CONTRACT_VERSIONS_V2 = {
  CreativeOrder: 'CreativeOrder.v2',
  ScriptApprovalReceipt: 'ScriptApprovalReceipt.v2',
  EditorApprovalReceipt: 'EditorApprovalReceipt.v2',
  PaidEffectIntent: 'PaidEffectIntent.v2',
  PaidEffectAttempt: 'PaidEffectAttempt.v2',
  VerifiedRenderReceipt: 'VerifiedRenderReceipt.v2',
} as const;

const NonEmptyString = z.string().trim().min(1);
const Sha256HexSchema = z.string().regex(/^[0-9a-f]{64}$/);
const CurrencyCodeSchema = z.string().regex(/^[A-Z]{3}$/);
const UtcTimestampSchema = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?Z$/)
  .refine((value) => !Number.isNaN(Date.parse(value)), 'invalid UTC timestamp');

function utcTimestampMicros(value: string): number {
  // Date.parse normalizes to milliseconds, which loses the six fractional
  // digits this sealed contract permits. Parse the fraction separately so an
  // update at .000001Z is correctly ordered after the whole second.
  const match = /^(.*?)(?:\.(\d{1,6}))?Z$/.exec(value);
  if (!match) return Number.NaN;
  const wholeMs = Date.parse(`${match[1]}Z`);
  if (Number.isNaN(wholeMs)) return Number.NaN;
  const micros = Number.parseInt((match[2] ?? '').padEnd(6, '0') || '0', 10);
  return wholeMs * 1000 + micros;
}

export const EffectKindSchema = z.enum([
  'visual', 'video', 'voiceover', 'music', 'sfx', 'render',
]);
export const PaidEffectStateSchema = z.enum([
  'PLANNED',
  'CLAIMED',
  'PROVIDER_STARTING',
  'PROVIDER_STARTED',
  'RECONCILE_REQUIRED',
  'SUCCEEDED',
  'FAILED_CONFIRMED',
  'NOT_STARTED_CONFIRMED',
]);

const EPHEMERAL_ORDER_FIELDS = new Set([
  'run_id', 'attempt_id', 'lease_id', 'provider_job_id',
]);

function isJsonValue(value: unknown): boolean {
  if (value === null || typeof value === 'string' || typeof value === 'boolean') return true;
  if (typeof value === 'number') return Number.isFinite(value);
  if (Array.isArray(value)) return value.every(isJsonValue);
  if (typeof value !== 'object') return false;
  return Object.entries(value as Record<string, unknown>).every(
    ([key, item]) => !EPHEMERAL_ORDER_FIELDS.has(key) && isJsonValue(item),
  );
}

const CanonicalOrderPayloadSchema = z
  .record(z.unknown())
  .refine((value) => Object.keys(value).length > 0, 'canonical_order_payload must not be empty')
  .refine(isJsonValue, 'canonical_order_payload must be JSON and exclude runtime identity');

const RevisionMapSchema = z
  .record(NonEmptyString)
  .refine((value) => Object.keys(value).length > 0, 'at least one revision is required')
  .refine(
    (value) => Object.keys(value).every((key) => key.trim().length > 0),
    'revision component names must be non-empty',
  );

function requireNonEmpty(value: string, field: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new Error(`${field} must be a non-empty string`);
  }
  return value.trim();
}

function requireDigest(value: string, field: string): string {
  if (!/^sha256:[0-9a-f]{64}$/.test(value)) {
    throw new Error(`${field} must be sha256:<64 lowercase hex>`);
  }
  return value;
}

function sha256Text(material: string): string {
  return createHash('sha256').update(material, 'utf8').digest('hex');
}

export function deriveCustomerOrderKeyV2(
  workspaceId: string,
  customerExternalOrderId: string,
): string {
  const workspace = requireNonEmpty(workspaceId, 'workspace_id');
  const external = requireNonEmpty(customerExternalOrderId, 'customer_external_order_id');
  return sha256Text(`${workspace}|${external}`);
}

export function deriveEffectKeyV2(
  customerOrderKey: string,
  approvedScriptDigest: string,
  effectKind: string,
  assetSlot: string,
): string {
  if (!/^[0-9a-f]{64}$/.test(customerOrderKey)) {
    throw new Error('customer_order_key must be 64 lowercase hex');
  }
  const script = requireDigest(approvedScriptDigest, 'approved_script_digest');
  const kind = requireNonEmpty(effectKind, 'effect_kind');
  const slot = requireNonEmpty(assetSlot, 'asset_slot');
  return sha256Text(`${customerOrderKey}|${script}|${kind}|${slot}`);
}

export function deriveEditorApprovalDigestV2(
  customerOrderKey: string,
  approvedScriptDigest: string,
  timelineDigest: string,
  mediaManifestDigest: string,
  renderPolicyDigest: string,
): string {
  if (!/^[0-9a-f]{64}$/.test(customerOrderKey)) {
    throw new Error('customer_order_key must be 64 lowercase hex');
  }
  return sha256Digest({
    customer_order_key: customerOrderKey,
    approved_script_digest: requireDigest(approvedScriptDigest, 'approved_script_digest'),
    timeline_digest: requireDigest(timelineDigest, 'timeline_digest'),
    media_manifest_digest: requireDigest(mediaManifestDigest, 'media_manifest_digest'),
    render_policy_digest: requireDigest(renderPolicyDigest, 'render_policy_digest'),
  });
}

export const CreativeOrderV2Schema = z
  .object({
    contract_version: z.literal('CreativeOrder.v2'),
    customer_order_key: Sha256HexSchema,
    workspace_id: NonEmptyString,
    account_id: NonEmptyString,
    brand_id: NonEmptyString,
    product_or_listing_id: NonEmptyString,
    customer_external_order_id: NonEmptyString,
    canonical_order_payload: CanonicalOrderPayloadSchema,
    canonical_order_digest: DigestSchema,
    created_at_utc: UtcTimestampSchema,
  })
  .strict()
  .superRefine((order, ctx) => {
    if (order.canonical_order_digest !== sha256Digest(order.canonical_order_payload)) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'canonical_order_digest mismatch' });
    }
    if (order.customer_order_key !== deriveCustomerOrderKeyV2(
      order.workspace_id, order.customer_external_order_id,
    )) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'customer_order_key mismatch' });
    }
  });
export type CreativeOrderV2 = z.infer<typeof CreativeOrderV2Schema>;

export const ScriptApprovalReceiptV2Schema = z
  .object({
    contract_version: z.literal('ScriptApprovalReceipt.v2'),
    approval_receipt_id: NonEmptyString,
    customer_order_key: Sha256HexSchema,
    workspace_id: NonEmptyString,
    approval_kind: z.literal('script'),
    approver_account_id: NonEmptyString,
    order_digest: DigestSchema,
    script_digest: DigestSchema,
    policy_digest: DigestSchema,
    approved_at_utc: UtcTimestampSchema,
    transaction_audit_id: NonEmptyString,
  })
  .strict();
export type ScriptApprovalReceiptV2 = z.infer<typeof ScriptApprovalReceiptV2Schema>;

export function scriptReceiptBindsOrder(
  receipt: ScriptApprovalReceiptV2,
  order: CreativeOrderV2,
  scriptDigest: string,
  policyDigest: string,
): boolean {
  return receipt.customer_order_key === order.customer_order_key
    && receipt.workspace_id === order.workspace_id
    && receipt.order_digest === order.canonical_order_digest
    && receipt.script_digest === scriptDigest
    && receipt.policy_digest === policyDigest;
}

export const EditorApprovalReceiptV2Schema = z
  .object({
    contract_version: z.literal('EditorApprovalReceipt.v2'),
    editor_approval_receipt_id: NonEmptyString,
    customer_order_key: Sha256HexSchema,
    workspace_id: NonEmptyString,
    editor_account_id: NonEmptyString,
    approved_script_digest: DigestSchema,
    timeline_digest: DigestSchema,
    media_manifest_digest: DigestSchema,
    render_policy_digest: DigestSchema,
    editor_approval_digest: DigestSchema,
    approved_at_utc: UtcTimestampSchema,
    transaction_audit_id: NonEmptyString,
  })
  .strict()
  .superRefine((receipt, ctx) => {
    const expected = deriveEditorApprovalDigestV2(
      receipt.customer_order_key,
      receipt.approved_script_digest,
      receipt.timeline_digest,
      receipt.media_manifest_digest,
      receipt.render_policy_digest,
    );
    if (receipt.editor_approval_digest !== expected) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'editor_approval_digest mismatch' });
    }
  });
export type EditorApprovalReceiptV2 = z.infer<typeof EditorApprovalReceiptV2Schema>;

export function editorReceiptBindsScriptApproval(
  editor: EditorApprovalReceiptV2,
  script: ScriptApprovalReceiptV2,
): boolean {
  return editor.customer_order_key === script.customer_order_key
    && editor.workspace_id === script.workspace_id
    && editor.approved_script_digest === script.script_digest
    && editor.render_policy_digest === script.policy_digest;
}

export const PaidEffectIntentV2Schema = z
  .object({
    contract_version: z.literal('PaidEffectIntent.v2'),
    effect_key: Sha256HexSchema,
    customer_order_key: Sha256HexSchema,
    workspace_id: NonEmptyString,
    approved_script_digest: DigestSchema,
    effect_kind: EffectKindSchema,
    asset_slot: NonEmptyString,
    request_digest: DigestSchema,
    spend_ceiling: z.number().finite().positive(),
    currency: CurrencyCodeSchema,
    created_at_utc: UtcTimestampSchema,
  })
  .strict()
  .superRefine((intent, ctx) => {
    const expected = deriveEffectKeyV2(
      intent.customer_order_key,
      intent.approved_script_digest,
      intent.effect_kind,
      intent.asset_slot,
    );
    if (intent.effect_key !== expected) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'effect_key mismatch' });
    }
  });
export type PaidEffectIntentV2 = z.infer<typeof PaidEffectIntentV2Schema>;

export const PaidEffectAttemptV2Schema = z
  .object({
    contract_version: z.literal('PaidEffectAttempt.v2'),
    effect_key: Sha256HexSchema,
    customer_order_key: Sha256HexSchema,
    workspace_id: NonEmptyString,
    approved_script_digest: DigestSchema,
    effect_kind: EffectKindSchema,
    asset_slot: NonEmptyString,
    attempt_id: NonEmptyString,
    attempt_number: z.number().int().positive(),
    provider: NonEmptyString,
    provider_idempotency_key: NonEmptyString,
    provider_job_id: NonEmptyString.nullable(),
    state: PaidEffectStateSchema,
    lease_owner: NonEmptyString.nullable(),
    lease_expires_at_utc: UtcTimestampSchema.nullable(),
    fencing_token: z.number().int().nonnegative(),
    request_digest: DigestSchema,
    spend_ceiling: z.number().finite().positive(),
    currency: CurrencyCodeSchema,
    response_digest: DigestSchema.nullable(),
    cost_currency: CurrencyCodeSchema.nullable(),
    cost_amount: z.number().finite().nonnegative().nullable(),
    last_reconciled_at_utc: UtcTimestampSchema.nullable(),
    created_at_utc: UtcTimestampSchema,
    updated_at_utc: UtcTimestampSchema,
  })
  .strict()
  .superRefine((attempt, ctx) => {
    const issue = (message: string) => ctx.addIssue({ code: z.ZodIssueCode.custom, message });
    const expected = deriveEffectKeyV2(
      attempt.customer_order_key,
      attempt.approved_script_digest,
      attempt.effect_kind,
      attempt.asset_slot,
    );
    if (attempt.effect_key !== expected) issue('effect_key mismatch');
    if (attempt.state !== 'PLANNED' && attempt.fencing_token < 1) issue('positive fencing_token required');
    if (['CLAIMED', 'PROVIDER_STARTING'].includes(attempt.state)
      && (attempt.lease_owner === null || attempt.lease_expires_at_utc === null)) {
      issue('active lease required');
    }
    if (['PROVIDER_STARTED', 'SUCCEEDED'].includes(attempt.state)
      && attempt.provider_job_id === null) {
      issue(`${attempt.state} requires provider_job_id`);
    }
    if (attempt.state === 'SUCCEEDED' && attempt.response_digest === null) {
      issue('SUCCEEDED requires response_digest');
    }
    if ((attempt.cost_currency === null) !== (attempt.cost_amount === null)) {
      issue('cost_currency and cost_amount must be present together');
    }
    if (attempt.cost_currency !== null && attempt.cost_currency !== attempt.currency) {
      issue('cost_currency must match authorized currency');
    }
    if (attempt.cost_amount !== null && attempt.cost_amount > attempt.spend_ceiling) {
      issue('cost_amount exceeds authorized spend_ceiling');
    }
    if (utcTimestampMicros(attempt.updated_at_utc) < utcTimestampMicros(attempt.created_at_utc)) {
      issue('updated_at cannot precede created_at');
    }
  });
export type PaidEffectAttemptV2 = z.infer<typeof PaidEffectAttemptV2Schema>;

export function paidEffectAttemptBindsIntent(
  attempt: PaidEffectAttemptV2,
  intent: PaidEffectIntentV2,
): boolean {
  return attempt.effect_key === intent.effect_key
    && attempt.customer_order_key === intent.customer_order_key
    && attempt.workspace_id === intent.workspace_id
    && attempt.approved_script_digest === intent.approved_script_digest
    && attempt.effect_kind === intent.effect_kind
    && attempt.asset_slot === intent.asset_slot
    && attempt.request_digest === intent.request_digest
    && attempt.spend_ceiling === intent.spend_ceiling
    && attempt.currency === intent.currency;
}

export function paidEffectAttemptAllowsNewAttempt(attempt: PaidEffectAttemptV2): boolean {
  return ['PLANNED', 'FAILED_CONFIRMED', 'NOT_STARTED_CONFIRMED'].includes(attempt.state);
}

export const VerifiedRenderReceiptV2Schema = z
  .object({
    contract_version: z.literal('VerifiedRenderReceipt.v2'),
    verified_render_receipt_id: NonEmptyString,
    customer_order_key: Sha256HexSchema,
    workspace_id: NonEmptyString,
    run_id: NonEmptyString,
    render_job_id: NonEmptyString,
    render_effect_key: Sha256HexSchema,
    editor_approval_digest: DigestSchema,
    output_url: z.string().url().refine((value) => value.startsWith('https://'), 'HTTPS required'),
    storage_key: NonEmptyString,
    output_sha256: Sha256HexSchema,
    output_bytes: z.number().int().positive(),
    duration_ms: z.number().int().positive(),
    video_codec: NonEmptyString,
    audio_codec: NonEmptyString,
    mechanical_checker_version: NonEmptyString,
    qa_checker_version: NonEmptyString,
    qa_verdict: z.literal('PASS'),
    qa_evidence_digest: DigestSchema,
    source_revisions: RevisionMapSchema,
    deployed_revisions: RevisionMapSchema,
    created_at_utc: UtcTimestampSchema,
    transaction_audit_id: NonEmptyString,
  })
  .strict();
export type VerifiedRenderReceiptV2 = z.infer<typeof VerifiedRenderReceiptV2Schema>;

export function verifiedReceiptBindsEditorApproval(
  receipt: VerifiedRenderReceiptV2,
  editor: EditorApprovalReceiptV2,
): boolean {
  return receipt.customer_order_key === editor.customer_order_key
    && receipt.workspace_id === editor.workspace_id
    && receipt.editor_approval_digest === editor.editor_approval_digest;
}

export function verifiedReceiptMatchesOutputBytes(
  receipt: VerifiedRenderReceiptV2,
  data: Uint8Array,
): boolean {
  return createHash('sha256').update(data).digest('hex') === receipt.output_sha256;
}

File: hiob_contracts/overnight_first_customer_v2.py
"""Strict first-customer truth-seam contracts (C2-C6, version 2).

Version 2 fixes the identity and validation weaknesses in the overnight v1
helpers:

* one external order has one stable key, independent of payload revisions,
  runs, and attempts; the stored canonical digest decides replay vs conflict;
* every approval/effect/render object is frozen and rejects extra/coerced data;
* paid work carries an explicit currency and spend ceiling;
* a verified render is representable only for QA PASS and an exact byte hash.
"""

from __future__ import annotations

import hashlib
import math
import re
from datetime import datetime
from typing import Annotated, Any, Literal, Mapping

from pydantic import (
    AfterValidator,
    BaseModel,
    ConfigDict,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)

from .factory.digest import is_digest, sha256_digest


FIRST_CUSTOMER_CONTRACT_VERSIONS_V2 = {
    "CreativeOrder": "CreativeOrder.v2",
    "ScriptApprovalReceipt": "ScriptApprovalReceipt.v2",
    "EditorApprovalReceipt": "EditorApprovalReceipt.v2",
    "PaidEffectIntent": "PaidEffectIntent.v2",
    "PaidEffectAttempt": "PaidEffectAttempt.v2",
    "VerifiedRenderReceipt": "VerifiedRenderReceipt.v2",
}

_FROZEN_STRICT = ConfigDict(frozen=True, extra="forbid", strict=True)
_KEY_RE = re.compile(r"^[0-9a-f]{64}$")
_UTC_RE = re.compile(
    r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?Z$"
)
_EPHEMERAL_ORDER_FIELDS = frozenset(
    {"run_id", "attempt_id", "lease_id", "provider_job_id"}
)


def _valid_utc(value: str) -> str:
    if not _UTC_RE.fullmatch(value):
        raise ValueError("timestamp must be an ISO-8601 UTC value ending in Z")
    try:
        parsed = datetime.fromisoformat(value[:-1] + "+00:00")
    except ValueError as exc:
        raise ValueError("timestamp is not a valid calendar value") from exc
    if parsed.utcoffset() is None or parsed.utcoffset().total_seconds() != 0:
        raise ValueError("timestamp must be UTC")
    return value


NonEmptyStr = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
DigestStr = Annotated[
    str, StringConstraints(pattern=r"^sha256:[0-9a-f]{64}$")
]
Sha256Hex = Annotated[str, StringConstraints(pattern=r"^[0-9a-f]{64}$")]
UtcTimestamp = Annotated[str, AfterValidator(_valid_utc)]
CurrencyCode = Annotated[str, StringConstraints(pattern=r"^[A-Z]{3}$")]
PositiveMoney = Annotated[float, Field(gt=0, allow_inf_nan=False)]
NonNegativeMoney = Annotated[float, Field(ge=0, allow_inf_nan=False)]
PositiveInt = Annotated[int, Field(gt=0)]
NonNegativeInt = Annotated[int, Field(ge=0)]

EffectKind = Literal["visual", "video", "voiceover", "music", "sfx", "render"]
PaidEffectState = Literal[
    "PLANNED",
    "CLAIMED",
    "PROVIDER_STARTING",
    "PROVIDER_STARTED",
    "RECONCILE_REQUIRED",
    "SUCCEEDED",
    "FAILED_CONFIRMED",
    "NOT_STARTED_CONFIRMED",
]


def _non_empty(value: str, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{field} must be a non-empty string")
    return value.strip()


def _digest(value: str, field: str) -> str:
    if not isinstance(value, str) or not is_digest(value):
        raise ValueError(f"{field} must be sha256:<64 lowercase hex>")
    return value


def _sha256_text(material: str) -> str:
    return hashlib.sha256(material.encode("utf-8")).hexdigest()


def derive_customer_order_key_v2(
    workspace_id: str,
    customer_external_order_id: str,
) -> str:
    """Stable business key: payload digest/run/attempt never participate.

    A persistence owner must compare ``canonical_order_digest`` separately:
    equal means idempotent replay; different means conflict for the same key.
    """

    workspace = _non_empty(workspace_id, "workspace_id")
    external = _non_empty(customer_external_order_id, "customer_external_order_id")
    return _sha256_text(f"{workspace}|{external}")


def derive_effect_key_v2(
    customer_order_key: str,
    approved_script_digest: str,
    effect_kind: str,
    asset_slot: str,
) -> str:
    """Stable paid-effect identity, excluding provider/run/attempt/lease data."""

    if not isinstance(customer_order_key, str) or not _KEY_RE.fullmatch(
        customer_order_key
    ):
        raise ValueError("customer_order_key must be 64 lowercase hex")
    script = _digest(approved_script_digest, "approved_script_digest")
    kind = _non_empty(effect_kind, "effect_kind")
    slot = _non_empty(asset_slot, "asset_slot")
    return _sha256_text(f"{customer_order_key}|{script}|{kind}|{slot}")


def derive_editor_approval_digest_v2(
    customer_order_key: str,
    approved_script_digest: str,
    timeline_digest: str,
    media_manifest_digest: str,
    render_policy_digest: str,
) -> str:
    """Content digest authorizing exactly one final-render input set."""

    if not isinstance(customer_order_key, str) or not _KEY_RE.fullmatch(
        customer_order_key
    ):
        raise ValueError("customer_order_key must be 64 lowercase hex")
    body = {
        "customer_order_key": customer_order_key,
        "approved_script_digest": _digest(
            approved_script_digest, "approved_script_digest"
        ),
        "timeline_digest": _digest(timeline_digest, "timeline_digest"),
        "media_manifest_digest": _digest(
            media_manifest_digest, "media_manifest_digest"
        ),
        "render_policy_digest": _digest(
            render_policy_digest, "render_policy_digest"
        ),
    }
    return sha256_digest(body)


def _validate_json_value(value: Any, path: str = "canonical_order_payload") -> None:
    if value is None or isinstance(value, (str, bool, int)):
        return
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValueError(f"{path} contains a non-finite number")
        return
    if isinstance(value, list):
        for index, item in enumerate(value):
            _validate_json_value(item, f"{path}[{index}]")
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                raise ValueError(f"{path} contains a non-string key")
            if key in _EPHEMERAL_ORDER_FIELDS:
                raise ValueError(f"{path} cannot contain ephemeral field {key!r}")
            _validate_json_value(item, f"{path}.{key}")
        return
    raise ValueError(f"{path} contains non-JSON value {type(value).__name__}")


def _validate_revisions(value: Mapping[str, str], field: str) -> Mapping[str, str]:
    if not value:
        raise ValueError(f"{field} must contain at least one revision")
    for name, revision in value.items():
        if not isinstance(name, str) or not name.strip():
            raise ValueError(f"{field} contains an empty component name")
        if not isinstance(revision, str) or not revision.strip():
            raise ValueError(f"{field}.{name} contains an empty revision")
    return value


class CreativeOrderV2(BaseModel):
    model_config = _FROZEN_STRICT

    contract_version: Literal["CreativeOrder.v2"]
    customer_order_key: Sha256Hex
    workspace_id: NonEmptyStr
    account_id: NonEmptyStr
    brand_id: NonEmptyStr
    product_or_listing_id: NonEmptyStr
    customer_external_order_id: NonEmptyStr
    canonical_order_payload: dict[str, Any]
    canonical_order_digest: DigestStr
    created_at_utc: UtcTimestamp

    @model_validator(mode="after")
    def _bind_identity(self) -> "CreativeOrderV2":
        if not self.canonical_order_payload:
            raise ValueError("canonical_order_payload must not be empty")
        _validate_json_value(self.canonical_order_payload)
        expected_digest = sha256_digest(self.canonical_order_payload)
        if self.canonical_order_digest != expected_digest:
            raise ValueError("canonical_order_digest does not match canonical_order_payload")
        expected_key = derive_customer_order_key_v2(
            self.workspace_id, self.customer_external_order_id
        )
        if self.customer_order_key != expected_key:
            raise ValueError("customer_order_key does not match external order identity")
        return self


class ScriptApprovalReceiptV2(BaseModel):
    model_config = _FROZEN_STRICT

    contract_version: Literal["ScriptApprovalReceipt.v2"]
    approval_receipt_id: NonEmptyStr
    customer_order_key: Sha256Hex
    workspace_id: NonEmptyStr
    approval_kind: Literal["script"]
    approver_account_id: NonEmptyStr
    order_digest: DigestStr
    script_digest: DigestStr
    policy_digest: DigestStr
    approved_at_utc: UtcTimestamp
    transaction_audit_id: NonEmptyStr

    def binds(
        self,
        order: CreativeOrderV2,
        script_digest: str,
        policy_digest: str,
    ) -> bool:
        return (
            self.customer_order_key == order.customer_order_key
            and self.workspace_id == order.workspace_id
            and self.order_digest == order.canonical_order_digest
            and self.script_digest == script_digest
            and self.policy_digest == policy_digest
        )


class EditorApprovalReceiptV2(BaseModel):
    model_config = _FROZEN_STRICT

    contract_version: Literal["EditorApprovalReceipt.v2"]
    editor_approval_receipt_id: NonEmptyStr
    customer_order_key: Sha256Hex
    workspace_id: NonEmptyStr
    editor_account_id: NonEmptyStr
    approved_script_digest: DigestStr
    timeline_digest: DigestStr
    media_manifest_digest: DigestStr
    render_policy_digest: DigestStr
    editor_approval_digest: DigestStr
    approved_at_utc: UtcTimestamp
    transaction_audit_id: NonEmptyStr

    @model_validator(mode="after")
    def _bind_digest(self) -> "EditorApprovalReceiptV2":
        expected = derive_editor_approval_digest_v2(
            self.customer_order_key,
            self.approved_script_digest,
            self.timeline_digest,
            self.media_manifest_digest,
            self.render_policy_digest,
        )
        if self.editor_approval_digest != expected:
            raise ValueError("editor_approval_digest does not match approved inputs")
        return self

    def binds(self, script_receipt: ScriptApprovalReceiptV2) -> bool:
        return (
            self.customer_order_key == script_receipt.customer_order_key
            and self.workspace_id == script_receipt.workspace_id
            and self.approved_script_digest == script_receipt.script_digest
            and self.render_policy_digest == script_receipt.policy_digest
        )


class PaidEffectIntentV2(BaseModel):
    model_config = _FROZEN_STRICT

    contract_version: Literal["PaidEffectIntent.v2"]
    effect_key: Sha256Hex
    customer_order_key: Sha256Hex
    workspace_id: NonEmptyStr
    approved_script_digest: DigestStr
    effect_kind: EffectKind
    asset_slot: NonEmptyStr
    request_digest: DigestStr
    spend_ceiling: PositiveMoney
    currency: CurrencyCode
    created_at_utc: UtcTimestamp

    @model_validator(mode="after")
    def _bind_effect_key(self) -> "PaidEffectIntentV2":
        expected = derive_effect_key_v2(
            self.customer_order_key,
            self.approved_script_digest,
            self.effect_kind,
            self.asset_slot,
        )
        if self.effect_key != expected:
            raise ValueError("effect_key does not match immutable paid-effect inputs")
        return self


class PaidEffectAttemptV2(BaseModel):
    model_config = _FROZEN_STRICT

    contract_version: Literal["PaidEffectAttempt.v2"]
    effect_key: Sha256Hex
    customer_order_key: Sha256Hex
    workspace_id: NonEmptyStr
    approved_script_digest: DigestStr
    effect_kind: EffectKind
    asset_slot: NonEmptyStr
    attempt_id: NonEmptyStr
    attempt_number: PositiveInt
    provider: NonEmptyStr
    provider_idempotency_key: NonEmptyStr
    provider_job_id: NonEmptyStr | None
    state: PaidEffectState
    lease_owner: NonEmptyStr | None
    lease_expires_at_utc: UtcTimestamp | None
    fencing_token: NonNegativeInt
    request_digest: DigestStr
    spend_ceiling: PositiveMoney
    currency: CurrencyCode
    response_digest: DigestStr | None
    cost_currency: CurrencyCode | None
    cost_amount: NonNegativeMoney | None
    last_reconciled_at_utc: UtcTimestamp | None
    created_at_utc: UtcTimestamp
    updated_at_utc: UtcTimestamp

    @model_validator(mode="after")
    def _bind_and_check_state(self) -> "PaidEffectAttemptV2":
        expected = derive_effect_key_v2(
            self.customer_order_key,
            self.approved_script_digest,
            self.effect_kind,
            self.asset_slot,
        )
        if self.effect_key != expected:
            raise ValueError("effect_key does not match immutable paid-effect inputs")
        if self.state != "PLANNED" and self.fencing_token < 1:
            raise ValueError("claimed or terminal attempts require a positive fencing_token")
        if self.state in {"CLAIMED", "PROVIDER_STARTING"} and (
            self.lease_owner is None or self.lease_expires_at_utc is None
        ):
            raise ValueError("claimed/provider-starting attempts require an active lease")
        if self.state in {"PROVIDER_STARTED", "SUCCEEDED"} and (
            self.provider_job_id is None
        ):
            raise ValueError(f"{self.state} requires provider_job_id")
        if self.state == "SUCCEEDED" and self.response_digest is None:
            raise ValueError("SUCCEEDED requires response_digest")
        if (self.cost_currency is None) != (self.cost_amount is None):
            raise ValueError("cost_currency and cost_amount must be present together")
        if self.cost_currency is not None and self.cost_currency != self.currency:
            raise ValueError("cost_currency must match authorized currency")
        if self.cost_amount is not None and self.cost_amount > self.spend_ceiling:
            raise ValueError("cost_amount exceeds authorized spend_ceiling")
        updated_at = _parse_utc_instant(self.updated_at_utc)
        created_at = _parse_utc_instant(self.created_at_utc)
        if updated_at is None or created_at is None:
            raise ValueError("attempt timestamps must be valid UTC instants")
        if updated_at < created_at:
            raise ValueError("updated_at_utc cannot precede created_at_utc")
        return self

    def binds(self, intent: PaidEffectIntentV2) -> bool:
        return (
            self.effect_key == intent.effect_key
            and self.customer_order_key == intent.customer_order_key
            and self.workspace_id == intent.workspace_id
            and self.approved_script_digest == intent.approved_script_digest
            and self.effect_kind == intent.effect_kind
            and self.asset_slot == intent.asset_slot
            and self.request_digest == intent.request_digest
            and self.spend_ceiling == intent.spend_ceiling
            and self.currency == intent.currency
        )

    def allows_new_attempt(self) -> bool:
        """Whether prior provider state proves another attempt is safe to start.

        Reconciliation is deliberately not retriable: a timeout can conceal a
        provider-side paid job even when no job id was returned.  Only pristine
        work or an explicitly terminal no/failed start may create an attempt.
        """
        return self.state in {"PLANNED", "FAILED_CONFIRMED", "NOT_STARTED_CONFIRMED"}


class VerifiedRenderReceiptV2(BaseModel):
    model_config = _FROZEN_STRICT

    contract_version: Literal["VerifiedRenderReceipt.v2"]
    verified_render_receipt_id: NonEmptyStr
    customer_order_key: Sha256Hex
    workspace_id: NonEmptyStr
    run_id: NonEmptyStr
    render_job_id: NonEmptyStr
    render_effect_key: Sha256Hex
    editor_approval_digest: DigestStr
    output_url: Annotated[str, StringConstraints(pattern=r"^https://\S+$")]
    storage_key: NonEmptyStr
    output_sha256: Sha256Hex
    output_bytes: PositiveInt
    duration_ms: PositiveInt
    video_codec: NonEmptyStr
    audio_codec: NonEmptyStr
    mechanical_checker_version: NonEmptyStr
    qa_checker_version: NonEmptyStr
    qa_verdict: Literal["PASS"]
    qa_evidence_digest: DigestStr
    source_revisions: dict[str, NonEmptyStr]
    deployed_revisions: dict[str, NonEmptyStr]
    created_at_utc: UtcTimestamp
    transaction_audit_id: NonEmptyStr

    @field_validator("source_revisions", "deployed_revisions")
    @classmethod
    def _non_empty_revisions(cls, value: Mapping[str, str], info: Any) -> Mapping[str, str]:
        return _validate_revisions(value, info.field_name)

    def binds(self, editor_receipt: EditorApprovalReceiptV2) -> bool:
        return (
            self.customer_order_key == editor_receipt.customer_order_key
            and self.workspace_id == editor_receipt.workspace_id
            and self.editor_approval_digest == editor_receipt.editor_approval_digest
        )

    def matches_output_bytes(self, data: bytes) -> bool:
        return hashlib.sha256(data).hexdigest() == self.output_sha256


def _parse_utc_instant(value: str) -> datetime | None:
    """Parse a contract timestamp for chronology, not lexicographic ordering."""
    try:
        return datetime.fromisoformat(value[:-1] + "+00:00")
    except (TypeError, ValueError):
        return None


def validate_creative_order_v2(obj: Mapping[str, Any]) -> CreativeOrderV2:
    return CreativeOrderV2.model_validate(obj)


def validate_script_approval_receipt_v2(
    obj: Mapping[str, Any],
) -> ScriptApprovalReceiptV2:
    return ScriptApprovalReceiptV2.model_validate(obj)


def validate_editor_approval_receipt_v2(
    obj: Mapping[str, Any],
) -> EditorApprovalReceiptV2:
    return EditorApprovalReceiptV2.model_validate(obj)


def validate_paid_effect_intent_v2(obj: Mapping[str, Any]) -> PaidEffectIntentV2:
    return PaidEffectIntentV2.model_validate(obj)


def validate_paid_effect_attempt_v2(obj: Mapping[str, Any]) -> PaidEffectAttemptV2:
    return PaidEffectAttemptV2.model_validate(obj)


def validate_verified_render_receipt_v2(
    obj: Mapping[str, Any],
) -> VerifiedRenderReceiptV2:
    return VerifiedRenderReceiptV2.model_validate(obj)


__all__ = [
    "FIRST_CUSTOMER_CONTRACT_VERSIONS_V2",
    "CreativeOrderV2",
    "ScriptApprovalReceiptV2",
    "EditorApprovalReceiptV2",
    "PaidEffectIntentV2",
    "PaidEffectAttemptV2",
    "VerifiedRenderReceiptV2",
    "EffectKind",
    "PaidEffectState",
    "derive_customer_order_key_v2",
    "derive_effect_key_v2",
    "derive_editor_approval_digest_v2",
    "validate_creative_order_v2",
    "validate_script_approval_receipt_v2",
    "validate_editor_approval_receipt_v2",
    "validate_paid_effect_intent_v2",
    "validate_paid_effect_attempt_v2",
    "validate_verified_render_receipt_v2",
]
